panic: empty structure "initial_lifecycle_hook." has entry "initial_lifecycle_hook.78676173.default_result"

### Terraform Version
0.8.5

### Affected Resource(s)
Please list the resources as a list, for example:
- aws_autoscaling_group

### Terraform Configuration Files
```hcl
resource "aws_autoscaling_group" "ecs_host_autoscaling_group" {
  # Required
  launch_configuration = "${aws_launch_configuration.host_launch_conf.name}"
  max_size = "${var.max_hosts}"
  min_size = "${var.min_hosts}"

  # Other
  name = "${aws_launch_configuration.host_launch_conf.name_prefix}autoscaling"
  vpc_zone_identifier = "${var.host_subnets}"
  initial_lifecycle_hook {
    name                   = "hook_ec2_instance_terminating"
    default_result         = "CONTINUE"
    heartbeat_timeout      = 3600
    lifecycle_transition   = "autoscaling:EC2_INSTANCE_TERMINATING"
    notification_target_arn = "${aws_sns_topic.ecs_host_autoscaling_group_lifecyclehook_notis.arn}"
    role_arn                = "${aws_iam_role.ecs_host_autoscaling_group_role.arn}"
  }

  depends_on = ["aws_launch_configuration.host_launch_conf","aws_sns_topic.ecs_host_autoscaling_group_lifecyclehook_notis"]
}
```

### Panic Output
panic: empty structure "initial_lifecycle_hook." has entry "initial_lifecycle_hook.78676173.default_result"
https://gist.github.com/pagemedias/b90b584ffc147be399437d53776332df

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. Create terraform config of ASG with `initial_lifecycle_hook` attribute
2. `terraform apply`
3. Remove `initial_lifecycle_hook` attribute from ASG
4. `terraform apply`

### References
Came across this while reporting... https://github.com/hashicorp/terraform/issues/11587

--- a/helper/schema/schema.go
+++ b/helper/schema/schema.go
@@ -919,6 +919,7 @@ func (m schemaMap) diffSet(
 	diff *terraform.InstanceDiff,
 	d *ResourceData,
 	all bool) error {
+
 	o, n, _, computedSet := d.diffChange(k)
 	if computedSet {
 		n = nil
@@ -1003,6 +1004,7 @@ func (m schemaMap) diffSet(
 		for _, code := range list {
 			switch t := schema.Elem.(type) {
 			case *Resource:
+				countDiff, cOk := diff.GetAttribute(k + ".#")
 				// This is a complex resource
 				for k2, schema := range t.Schema {
 					subK := fmt.Sprintf("%s.%s.%s", k, code, k2)
@@ -1010,7 +1012,17 @@ func (m schemaMap) diffSet(
 					if err != nil {
 						return err
 					}
+
+					// If parent set is being removed
+					// remove all subfields which were missed by the diff func
+					// We process these separately because type-specific diff functions
+					// lack the context (hierarchy of fields)
+					subKeyIsCount := strings.HasSuffix(subK, ".#")
+					if cOk && countDiff.New == "0" && !subKeyIsCount {
+						m.markAsRemoved(subK, schema, diff)
+					}
 				}
+
 			case *Schema:
 				// Copy the schema so that we can set Computed/ForceNew from
 				// the parent schema (the TypeSet).

--- a/helper/schema/schema_test.go
+++ b/helper/schema/schema_test.go
@@ -2985,6 +2985,95 @@ func TestSchemaMap_Diff(t *testing.T) {
 				},
 			},
 		},
+
+		{
+			Name: "Removal of TypeSet should cause computed fields to be removed",
+			Schema: map[string]*Schema{
+				"type_set": &Schema{
+					Type:     TypeSet,
+					Optional: true,
+					Elem: &Resource{
+						Schema: map[string]*Schema{
+							"name": &Schema{
+								Type:     TypeString,
+								Optional: true,
+							},
+							"required": &Schema{
+								Type:     TypeString,
+								Required: true,
+							},
+							"value": &Schema{
+								Type:     TypeInt,
+								Optional: true,
+							},
+							"required_value": &Schema{
+								Type:     TypeInt,
+								Required: true,
+							},
+							"computed_value": &Schema{
+								Type:     TypeString,
+								Optional: true,
+								Computed: true,
+							},
+						},
+					},
+					Set: func(i interface{}) int {
+						if i != nil {
+							return 12345
+						}
+						return 0
+					},
+				},
+			},
+
+			State: &terraform.InstanceState{
+				Attributes: map[string]string{
+					"type_set.#":                    "1",
+					"type_set.12345.name":           "Name",
+					"type_set.12345.required":       "Required",
+					"type_set.12345.value":          "0",
+					"type_set.12345.required_value": "5",
+					"type_set.12345.computed_value": "COMPUTED",
+				},
+			},
+
+			Config: map[string]interface{}{
+				"type_set": []interface{}{},
+			},
+
+			Diff: &terraform.InstanceDiff{
+				Attributes: map[string]*terraform.ResourceAttrDiff{
+					"type_set.#": &terraform.ResourceAttrDiff{
+						Old:        "1",
+						New:        "0",
+						NewRemoved: false,
+					},
+					"type_set.12345.name": &terraform.ResourceAttrDiff{
+						Old:        "Name",
+						New:        "",
+						NewRemoved: true,
+					},
+					"type_set.12345.required": &terraform.ResourceAttrDiff{
+						Old:        "Required",
+						New:        "",
+						NewRemoved: true,
+					},
+					"type_set.12345.value": &terraform.ResourceAttrDiff{
+						Old:        "0",
+						New:        "0",
+						NewRemoved: true,
+					},
+					"type_set.12345.required_value": &terraform.ResourceAttrDiff{
+						Old:        "5",
+						New:        "0",
+						NewRemoved: true,
+					},
+					"type_set.12345.computed_value": &terraform.ResourceAttrDiff{
+						NewRemoved: true,
+					},
+				},
+			},
+		},
 	}
 
 	for i, tc := range cases {

